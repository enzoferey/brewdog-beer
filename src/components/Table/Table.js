import React from "react";
import PropTypes from "prop-types";
import style from "./table.scss";

import theme from "theme";

import TableRow from "components/TableRow";
import Hop from "components/Hop";
import Malt from "components/Malt";
import Method from "components/Method";

const getElement = type => {
  switch (type) {
    case "hop":
      return Hop;
    case "malt":
      return Malt;
    case "method":
      return Method;
    default:
      return DefaultItem;
  }
};

export const getWaiting = (type, time, startDone, middleDone) => {
  return (
    type === "hop" &&
    ((time === "middle" && !startDone) || (time === "end" && !middleDone))
  );
};

export const getDuration = (type, temperatures) => {
  if (type === "method") {
    return temperatures.reduce(
      (prev, temperature) => prev + temperature.duration,
      0
    );
  }

  return 0;
};

// Just a placeholder in case somethind else is passed
const DefaultItem = props => <p>{JSON.stringify(props)}</p>;

const Table = ({
  title,
  type,
  rows,
  startDone,
  middleDone,
  setDone,
  gradient,
}) => {
  const Item = getElement(type);

  return (
    <div className={style.main}>
      <h2 className={style.title}>{title}</h2>
      <div className={style.table}>
        {rows.map((row, index) => {
          const waiting = getWaiting(type, row.add, startDone, middleDone);
          const duration = getDuration(type, row.mash_temp);
          const gradientButton = waiting ? theme.gradientGrey : gradient;
          return (
            <TableRow
              key={`${title}-row-${index}`}
              item={<Item {...row} />}
              done={row.done}
              duration={duration}
              waiting={waiting}
              callback={() => {
                setDone(index);
              }}
              gradientButton={gradientButton}
            />
          );
        })}
      </div>
    </div>
  );
};

Table.propTypes = {
  title: PropTypes.string,
  type: PropTypes.oneOf(["hop", "malt", "method"]),
  rows: PropTypes.arrayOf(PropTypes.object),
  startDone: PropTypes.bool,
  middleDone: PropTypes.bool,
  setDone: PropTypes.func,
  gradient: PropTypes.string,
};

export default Table;
