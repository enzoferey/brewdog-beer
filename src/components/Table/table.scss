.main {
  margin-top: 55px;
}

.title {
  font-size: 30px;
  margin-bottom: 20px;
  font-weight: 400;
}
