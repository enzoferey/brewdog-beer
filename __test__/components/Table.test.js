import React from "react";
import Table from "components/Table";
import { getWaiting, getDuration } from "components/Table/Table";

const props = {
  title: "My title",
  type: "hop",
  rows: mocks.beers[0].ingredients.hops,
  startDone: false,
  middleDone: false,
  setDone: jest.fn(),
  gradient: "white",
};

test("renders correctly", () => {
  testSnapshot(<Table {...props} />);
});

const checkWaiting = (type, time, startDone, middleDone, result) => {
  expect(getWaiting(type, time, startDone, middleDone)).toBe(result);
};

describe("getWaiting()", () => {
  describe("when element is not a hop", () => {
    it("should return false", () => {
      checkWaiting("whatever", "start", true, true, false);
    });
  });
  describe("when element is a hop", () => {
    describe('when time field is "start"', () => {
      it("should return false", () => {
        checkWaiting("hop", "start", true, true, false);
      });
    });
    describe('when time field is "middle"', () => {
      it("should return true if startDone=false", () => {
        checkWaiting("hop", "middle", false, false, true);
      });
      it("should return false if startDone=true", () => {
        checkWaiting("hop", "middle", true, false, false);
      });
    });
    describe('when time field is "end"', () => {
      it("should return true if middleDone=false", () => {
        checkWaiting("hop", "end", true, false, true);
      });
      it("should return false if middleDone=true", () => {
        checkWaiting("hop", "end", true, true, false);
      });
    });
  });
});

const checkDuration = (type, temperatures, result) => {
  expect(getDuration(type, temperatures)).toBe(result);
};

describe("getDuration()", () => {
  describe("when element is not a method", () => {
    it("should return 0", () => {
      checkDuration("whatever", [], 0);
    });
  });
  describe("when element is a method", () => {
    it("should return the sum of the temperatures' durations, even when null values", () => {
      checkDuration("method", mocks.temperaturesSingle, 10);
      checkDuration("method", mocks.temperaturesNull, 10);
      checkDuration("method", mocks.temperaturesMultiple, 15);
    });
  });
});
