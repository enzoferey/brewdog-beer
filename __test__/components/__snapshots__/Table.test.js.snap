// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <h2>
    My title
  </h2>
  <div>
    <div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <span>
          Fuggles
        </span>
      </div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <img
          alt=""
          src="weight.png"
        />
        <span>
          25 grams
        </span>
      </div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <img
          alt=""
          src="time.png"
        />
        <span>
          start
        </span>
      </div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <img
          alt=""
          src="label.png"
        />
        <span>
          bitter
        </span>
      </div>
      <div>
        <div
          className="undefined "
          onClick={[Function]}
          style={
            Object {
              "background": "white",
            }
          }
        >
          <p>
            Add
          </p>
          <div
            className="undefined "
            style={
              Object {
                "width": "NaN%",
              }
            }
          />
        </div>
      </div>
    </div>
    <div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <span>
          First Gold
        </span>
      </div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <img
          alt=""
          src="weight.png"
        />
        <span>
          25 grams
        </span>
      </div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <img
          alt=""
          src="time.png"
        />
        <span>
          start
        </span>
      </div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <img
          alt=""
          src="label.png"
        />
        <span>
          bitter
        </span>
      </div>
      <div>
        <div
          className="undefined "
          onClick={[Function]}
          style={
            Object {
              "background": "white",
            }
          }
        >
          <p>
            Add
          </p>
          <div
            className="undefined "
            style={
              Object {
                "width": "NaN%",
              }
            }
          />
        </div>
      </div>
    </div>
    <div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <span>
          Fuggles
        </span>
      </div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <img
          alt=""
          src="weight.png"
        />
        <span>
          37.5 grams
        </span>
      </div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <img
          alt=""
          src="time.png"
        />
        <span>
          middle
        </span>
      </div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <img
          alt=""
          src="label.png"
        />
        <span>
          flavour
        </span>
      </div>
      <div>
        <div
          className="undefined undefined"
          onClick={[Function]}
          style={
            Object {
              "background": undefined,
            }
          }
        >
          <p>
            Add prev
          </p>
          <div
            className="undefined "
            style={
              Object {
                "width": "NaN%",
              }
            }
          />
        </div>
      </div>
    </div>
    <div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <span>
          First Gold
        </span>
      </div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <img
          alt=""
          src="weight.png"
        />
        <span>
          37.5 grams
        </span>
      </div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <img
          alt=""
          src="time.png"
        />
        <span>
          middle
        </span>
      </div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <img
          alt=""
          src="label.png"
        />
        <span>
          flavour
        </span>
      </div>
      <div>
        <div
          className="undefined undefined"
          onClick={[Function]}
          style={
            Object {
              "background": undefined,
            }
          }
        >
          <p>
            Add prev
          </p>
          <div
            className="undefined "
            style={
              Object {
                "width": "NaN%",
              }
            }
          />
        </div>
      </div>
    </div>
    <div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <span>
          Cascade
        </span>
      </div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <img
          alt=""
          src="weight.png"
        />
        <span>
          37.5 grams
        </span>
      </div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <img
          alt=""
          src="time.png"
        />
        <span>
          end
        </span>
      </div>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <img
          alt=""
          src="label.png"
        />
        <span>
          flavour
        </span>
      </div>
      <div>
        <div
          className="undefined undefined"
          onClick={[Function]}
          style={
            Object {
              "background": undefined,
            }
          }
        >
          <p>
            Add prev
          </p>
          <div
            className="undefined "
            style={
              Object {
                "width": "NaN%",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
